### DISCUSSION TITLE
Struggles to converge on new MOOSE app

### DISCUSSION THREAD

**Discussion by wehrad** [idaholab/moose#26651] [Q&A Getting Started]
Hi all,

First and foremost, thank you a lot for your precious support and help in the past!

I'm currently working on a MOOSE app for ice viscous deformation. After many trials in different setups with different solvers and initial conditions I just can't get my solution to converge.

[Here](https://github.com/AdrienWehrle/ice-worm/blob/devel/inputs/viscous_deformation/boundary_coupling/mooseice_vd_flow_transient_sb_bc_v2.i) is the main input file I'm currently using. The mesh is [here](https://github.com/AdrienWehrle/ice-worm/blob/devel/meshes/mesh_ac_lowres.e) and the two main Material objects are [IceMaterial_u2](https://github.com/AdrienWehrle/ice-worm/blob/devel/src/materials/IceMaterial_u2.C) (with ice viscosity based on Glen's flow law relating the rate of deformation) and [SedimentMaterial_u2](https://github.com/AdrienWehrle/ice-worm/blob/devel/src/materials/SedimentMaterial_u2.C) (constant viscosity for a start).

I'm wondering if for instance the executioner used [here](https://github.com/AdrienWehrle/ice-worm/blob/80a7b7c663995ec49302bd3a585ef60154f49a11/inputs/viscous_deformation/boundary_coupling/mooseice_vd_flow_transient_sb_bc_v2.i#L279-L324) is not appropriate to the situation and should be changed?

The setup consists in a glacier in contact with the ocean on its downstream boundary, sliding over a sediment layer and having a
Dirichlet boundary condition at the upstream boundary to account for the influx of ice further up glacier (1000 m.y-1 in the input file shared above). For such cases, we usually work with MPa a so that stresses are of the order of 1 MPa. I think my dimension analysis should be correct but maybe the convergence issue comes from this and I just can't find it out...! 

I'm using MOOSE through another fork that is up to date (and all tests pass):
```
Framework Information:
MOOSE Version:           git commit cee3f610f9 on 2024-01-25
LibMesh Version:         
PETSc Version:           3.20.1
SLEPc Version:           3.20.0
```

Here an example of run that never converges:
```
Time Step 0, time = 0

Time Step 1, time = 0.001, dt = 0.001

Performing automatic scaling calculation

 0 Nonlinear |R| = 9.856816e+08
Currently Executing    Finished Computing Jacobian                                                          [  1.95 s] [  557 MB]
      0 Linear |R| = 9.856816e+08
      1 Linear |R| = 3.098366e+03
 1 Nonlinear |R| = 9.856815e+08
      0 Linear |R| = 9.856815e+08
      1 Linear |R| = 3.704444e+02
 2 Nonlinear |R| = 9.852308e+08
```

Thanks a lot in advance for your help!

**Comment by GiudGiud** [comment]
Hello

Looks like you definitely have to be careful with the units here. I see year and MPa in that post.
Could you try a separate input file in s, m, Pa, ... (you can scale the mesh to go to meters) and make sure you get exactly the same results?

If that does not help, then can you please report on a few of the troubleshooting solves steps?
https://mooseframework.inl.gov/moose/application_usage/failed_solves.html

condition number for a tiny problem would be nice. Mesh diagnostics is sometimes helpful

Guillaume

**Reply by wehrad** [reply]
Hi Guillaume,

thanks a lot for your help! 

Unfortunately both cases in `Pa.s` and `MPa.a` don't converge at the moment so I can't even compare them to make sure they're similar...

Following the advice on failed solves I've been working on [a simple gold case](https://github.com/AdrienWehrle/ice-worm/blob/devel/inputs/viscous_deformation/boundary_coupling/mooseice_vd_flow_transient_gold.i) of an [ice slab](https://github.com/AdrienWehrle/ice-worm/blob/devel/meshes/mesh_gold_lowerres.e) flowing on a sloped surface with null Dirichlet BC at the base and on the sides, and periodic BCs on the upstream and downstream faces of the ice slab. I realized I still couldn't reach convergence even on this very simple case. It hence looks like it's the convergence of the ice rheology itself (the physics behind has been proof-checked several times) that I'm struggling with more than the different setups I'm using.

From there, I've been trying to:

- set ICs closer to the solution (no significant improvement on convergence)
- add a strong diffusion term at the beginning of the simulation (converges well at the beginning, slowly increasing velocity in the ice slab, then converging harder and harder as the diffusion term fades, until it wouldn't converge anymore)
- add mesh adaptivity in case the simple mesh needed refinement to converge (eg close to the base; but no significant improvement either)
- set a higher initial tolerance (both `nl_rel_tol` and `nl_abs_tol` are set to `1e-4` as you can see on the input file shared through the hyperlink above; but that didn't help either)

What I find interesting is that when I set an initial ice viscosity of 0.3 MPa.a all over the ice slab (which could be a value we expect) then velocity and pressure are solved very easily. Once I turn off the viscosity dependence on strain rate (at the second time step), no convergence can be reached anymore...

Thank you a lot for your precious help on this!

**Reply by GiudGiud** [reply]
Thank you for investigating all this. I'll take a look next week unless @lindsayad or @grmnptr get to it first on this FE flow problem

**Reply by wehrad** [reply]
Thank you a lot, @GiudGiud !

The past days I've been restarting the development of the app from the [Navier Stokes velocity inlet BC input file](https://github.com/idaholab/moose/blob/a613f4c4c58965ee0cbd5219fe0f3f02b0c9fd07/modules/navier_stokes/test/tests/finite_element/ins/velocity_channel/velocity_inletBC_no_parts.i#L4) used in moose for testing, simulating flow through a channel with incompressible Navier Stokes.

I've been scaling up the geometry from the ~1m scale to a `10*2*0.5 km` channel with constant viscosity to start with. As I'm dealing with ice, I have a viscosity of the order of `10^12 Pa.s`, motivating the switch to `MPa.a`, which I did again in this input file (see [here](https://github.com/AdrienWehrle/ice-worm/blob/9c2d368365a8d96aa669b6202c5396f5e595deed/inputs/viscous_deformation/boundary_coupling/mooseice_from_velocity_inletBC.i#L139) together with the `1e-6` factor on gravity). Here already, convergence was a bit hard to get to, but using a pretty high viscosity, I eventually managed. 

Unfortunately, as soon as I incorporate the viscosity dependency on strain rate (Glen's flow law for ice) through an [IceMaterial](https://github.com/AdrienWehrle/ice-worm/blob/devel/src/materials/IceMaterial_u2.C), I can't reach convergence in any case anymore...

Here is my [final input file](https://github.com/AdrienWehrle/ice-worm/blob/devel/inputs/viscous_deformation/boundary_coupling/mooseice_from_velocity_inletBC.i) where the solve starts to fail, which is simpler to reproduce than the file I shared above as e.g. I use the mesh generator.

Thank you a lot for your precious help on this!

**Reply by GiudGiud** [reply]
I dont see any derivatives in this IceMaterial. Do the equations not involve any drho/dT(P) or dmu/dT(P) in the Jacobian?
Could you try an AD version of this IceMaterial ? (replace Real by ADReal everywhere in the material, and also in the places where you use those material properties)

**Reply by wehrad** [reply]
> I dont see any derivatives in this IceMaterial. Do the equations not involve any drho/dT(P) or dmu/dT(P) in the Jacobian?

The density is constant, and the viscosity I implemented is based on Equation 6 [here](https://tc.copernicus.org/articles/12/721/2018/). 

> Could you try an AD version of this IceMaterial ? 

I'll try this right now!

**Reply by wehrad** [reply]
> Could you try an AD version of this IceMaterial ?

Converting all `Real` on my material source and header to `ADReal`, I get the following error

```
*** ERROR ***
The following error occurred in the object "mass", of type "INSMass".

The requested non-AD material property 'mu' of type 'double'
is already retrieved or declared as a non-AD property of type 'MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, unsigned long, MetaPhysicL::NWrapper<64ul>>, true>'.
```

does that mean I should make some changes in INSMass? I always feel like when I ask myself this question the answer is rather that something needs to change in my material or input file, but in this case I can't find the issue...!

**Reply by GiudGiud** [reply]
so rho is incompressible. No derivatives there is normal,
The only other material property is then mu.

mu does not depend on pressure, so you dont get an extra term in the Jacobian from that. But it depends on velocities, so when forming the Jacobian for the momentum equation I would expect something to show up.
Using AD is the easiest way to get that working.
The only thing to be careful about is not losing the derivatives anywhere in the computaiton by using 'Real' instead of 'ADReal'

**Reply by GiudGiud** [reply]
INSMass retrieves mu? That's weird. Anyhow, yes unfortunately the kernels also have to be updated to use AD. 

I would make a new input file that uses INSAD instead of INS. 
If you wanted to take into account the derivatives of viscosity wrt velocity you would need to modify the INSMomentum kernel too (add work in computeQpJacobian)

**Reply by wehrad** [reply]
Thanks a lot for all these important points. So I think I understand I need to modify INSMomentum to work with AD if I want to use it with INSAD, is that correct?

**Reply by GiudGiud** [reply]
> So I think I understand I need to modify INSMomentum to work with AD if I want to use it with INSAD, is that correct?

no if you want to work with AD, starting from INSAD is easier. Maybe look at example inputs for INSAD though, there's some differences
Modifying INS is going to be trickier

**Reply by wehrad** [reply]
Thanks a lot, Guillaume. Following your advice I've been using [a test input file](https://github.com/idaholab/moose/blob/a613f4c4c58965ee0cbd5219fe0f3f02b0c9fd07/modules/navier_stokes/test/tests/finite_element/ins/lid_driven/ad_lid_driven.i#L65) with INSAD and from there created [a new input file](https://github.com/AdrienWehrle/ice-worm/blob/devel/inputs/viscous_deformation/boundary_coupling/mooseice_from_ad_lid_driven.i) for my application, using INSAD instead of INS.

I now run into the following error which I struggle debugging:

```
*** ERROR ***
The following error occurred in the object "MOOSE Problem", of type "FEProblem".

Unable to find user object with name 'ins_ad_object_tracker'
```

which might be somehow linked to https://github.com/idaholab/moose/issues/15713?

**Reply by GiudGiud** [reply]
Hello

No that issue was solved, and it was a threading problem so it would not be an issue for you now since I imagine you are not using threading. 

You are missing the material for INSAD
```
  [ins_mat]
    type = INSAD3Eqn
    velocity = velocity
    pressure = p
    temperature = T
  []
```
use a constant field for temperature since you are not solving for it

**Reply by wehrad** [reply]
Thank you! INSAD is now working fine now for me with a generic constant material although it doesn't look like it's helping a lot, but maybe it will on my ice material.
Now I run back to an issue that I had before with variable type when I try to use my custom ice material, apparently coming from the way I declare rho in my material [here](https://github.com/AdrienWehrle/ice-worm/blob/devel/src/materials/IceMaterial_u2_ad.C#L47). As if I wasn't overwriting rho properly. Thank you a lot for all your support, doing my best to debug myself as much as possible but I often get stuck.

```
*** ERROR ***
The following error occurred in the object "ice", of type "IceMaterial_u2_ad".

The declared AD material property 'rho' of type 'MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, unsigned long, MetaPhysicL::NWrapper<64ul>>, true>'
is already retrieved or declared as a AD property of type 'double'.
```

**Reply by GiudGiud** [reply]
You are declaring rho twice in two materials. That should be the problem

**Reply by wehrad** [reply]
Yes it looks like INSAD3Eqn wants to retrieve/declare it again after I first set it with my ice material.

**Reply by GiudGiud** [reply]
That s annoying. Maybe just remove it from your material. 

In the FV version of this material we have a Boolean to define or not the density material property

**Reply by wehrad** [reply]
> That s annoying. Maybe just remove it from your material.

It's unfortunately the same issue with mu and I need to compute my variable mu in my material..

**Reply by GiudGiud** [reply]
I just checked the INSAD material. It is not declaring these properties. It is retrieving them. 

So something in the declaration and the retrieval is not matching.

**Reply by GiudGiud** [reply]
Ah I see it now. 

You call in your material 'declare AD Property < ADReal '

But it should be 

'declare AD Property < Real '
